### PR TITLE
Add ProjectDirectoryProvider

### PR DESCRIPTION
**Pull request by rita-sonarsource**
Prep for #2195

### CODE REVIEW

File: src/TypeScript/TsConfig/TsConfigsLocator.cs [hunk 1]
```diff
@@ -93,41 +73,5 @@ public IReadOnlyList<string> Locate(string sourceFilePath)
...
-                    "Failed to retrieve VSPROPID_SolutionDirectory");
-
-                return (string) solutionDirectory;
-            }
-
-            var vsProject = vsHierarchyLocator.GetVsHierarchyForFile(sourceFilePath);
-
-            if (vsProject == null)
-            {
-                logger.WriteLine(Resources.ERR_NoVsProject, sourceFilePath);
```
Comment (rita-sonarsource) [inline src/TypeScript/TsConfig/TsConfigsLocator.cs:119]: extracted as-is minus this log output

File: src/Infrastructure.VS/ProjectDirectoryProvider.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        public ProjectDirectoryProvider([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IVsHierarchyLocator vsHierarchyLocator)
+        {
+            vsSolution = serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution;
+            this.vsHierarchyLocator = vsHierarchyLocator;
+        }
+
+        public string GetProjectDirectory(string sourceFilePath)
+        {
+            // "Open as Folder" was introduced in VS2017, so in VS2015 the hr result will be `E_NOTIMPL` and `isOpenAsFolder` will be null.
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS/ProjectDirectoryProvider.cs:61]: Looking at this again, we're making a fairly big assumption, namely that in the "open as folder" case the supplied file path is under the folder root. That's true for the way we call it currently, but it looks odd now the interface has been separated out.

Effectively, the behaviour in the "open as folder" case is that the caller can pass in any garbage and the function will return the root folder path.
I think we should either document this behaviour on the interface, or add a check to the code.

**Review by duncanp-sonar — COMMENTED** [review]
The changes look good.

However, I'm not clear how we'll use this in CMake scenario. Which file name are we going to pass in to `GetProjectDirectory`?

File: src/Infrastructure.VS/ProjectDirectoryProvider.cs (hunk 2 above)
Comment (rita-sonarsource) [reply src/Infrastructure.VS/ProjectDirectoryProvider.cs:61]: This is actually exactly the design problem I'm 'struggling' with right now.
Eventually the higher caller would call something like
`CompilationOptions Get(string analyzedFile)` --> and that in turn would get the compilation database, get the entry and parse it.
So in a way we do always start with a file, but strictly speaking, I'm writing a class now `CompilationDatabase Locate()` that absolutely does not need a file path.

**Pull request closed without merging**